### PR TITLE
[Feature] Introducing image compression using imagemin

### PR DESCRIPTION
**Pull request by ahmadalfy**
## Motivation

Static assets should always be compressed to improve loading speed and enhance user experience. The static images in the user folder was 4.4MB. After using imagemin the total size of the image was reduced to nearly the half.

Also this PR include replacing the images in the example website with the optimized ones to improve the download speed for future users.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Related Issues

https://github.com/facebook/Docusaurus/issues/568

### CODE REVIEW

File: lib/server/generate.js [hunk 1]
```diff
@@ -399,6 +404,24 @@ async function execute() {
       }
 
       fs.writeFileSync(mainCss, cssContent);
+    } else if (normalizedFile.match(/\.png$|.jpg$|.svg$|.gif$/)) {
+      console.log(normalizedFile);
```
Comment (yangshun) [inline lib/server/generate.js]: `console.log` should be removed.

File: lib/server/generate.js [hunk 2]
```diff
@@ -399,6 +404,24 @@ async function execute() {
...
+      let parts = normalizedFile.split(sep + 'static' + sep);
+      let targetDirectory = join(
```
Comment (yangshun) [inline lib/server/generate.js]: Use `const`.

**Comment by ahmadalfy** [comment]
For some reason `imagemin-svgo` wasn't added to `package.json`. Let me fix that

File: lib/server/generate.js (hunk 1 above)
Comment (ahmadalfy) [reply lib/server/generate.js]: Opps!

File: lib/server/generate.js (hunk 2 above)
Comment (ahmadalfy) [reply lib/server/generate.js]: Was following the same convention used on the next block, should I update that one as well?

Comment (yangshun) [reply lib/server/generate.js]: Sure!

**Comment by JoelMarcey** [comment]
Re: a flag, is the image quality that diminished that the average person would notice on a website? If so, a flag makes more sense to me. If not, then a flag is probably still good, but maybe less of a mandatory thing.

**Comment by ahmadalfy** [comment]
This is a lossless compression, image quality is not affected as it is only removing unnecessary meta data like EXIF info and others.

**Comment by ahmadalfy** [comment]
OK does the flag name `--no-compression` sound good ?

**Comment by yangshun** [comment]
Let's call it `--skip-image-compression`. We'll have to document it too 👍

**Comment by JoelMarcey** [comment]
Actually -- a name like `no-image-compression` would be more descriptive, I think. Otherwise people might ask what we are compressing.

**Comment by JoelMarcey** [comment]
`--skip-image-compression` is good too.

**Comment by yangshun** [comment]
I agree. Because we might potentially add flags for the CSS compression (and other types of assets).

**Comment by ahmadalfy** [comment]
Perfect, will do

**Comment by ahmadalfy** [comment]
Does this sound good:

```md
This will generate a `build` folder inside the `website` directory containing the `.html`
files from all of your docs and other pages included in `pages`. The build command will
also compress all the images you place in your assets directory unless you supply it with
the flag `--skip-image-compression`.
```

**Comment by yangshun** [comment]
No, that flag should be documented on this page - https://docusaurus.io/docs/en/commands.html#docusaurus-build

**Comment by ahmadalfy** [comment]
Flag updated and added to the docs.

**Comment by yangshun** [comment]
@ahmadalfy CI and the preview build is failing and I'm not sure why. Could you try to build locally and make sure your code is prettified?

**Comment by ahmadalfy** [comment]
I ran prettier and introduced a change, build working though through

```
cd website
yarn build
```

Should work now :)

**Comment by yangshun** [comment]
Awesome! Thank you for the PR! Such a huge win for us 😄

**Comment by ahmadalfy** [comment]
Thanks for the welcoming spirit 💘

**Pull request merged**